File: vscode-plugin/src/telemetry/privacy.ts
/**
 * Privacy helpers for VS Code telemetry (Slice 9.13).
 */

import { APPROVED_FIELD_NAMES } from "./events";

const FORBIDDEN_NAME_FRAGMENTS = [
  "workspace",
  "repository",
  "document",
  "path",
  "finding",
  "evidence",
  "prompt",
  "response",
  "credential",
  "password",
  "token",
  "secret",
  "api_key",
  "installation",
  "machine",
  "argv",
  "model",
  "provider",
  "cost",
] as const;

const UNSAFE_VALUE_RE =
  /(\/Users\/|\/home\/|[A-Za-z]:\\|https?:\/\/|file:\/\/|-----BEGIN|password=|secret=)/i;

export function isApprovedFieldName(name: string): boolean {
  return (APPROVED_FIELD_NAMES as readonly string[]).includes(name);
}

export function isForbiddenFieldName(name: string): boolean {
  const lower = name.toLowerCase();
  if ((APPROVED_FIELD_NAMES as readonly string[]).includes(lower)) {
    return false;
  }
  return FORBIDDEN_NAME_FRAGMENTS.some((frag) => lower.includes(frag));
}

export function looksLikeUnsafeValue(value: unknown): boolean {
  if (typeof value !== "string") {
    return false;
  }
  if (value.includes("\n") || value.includes("\r")) {
    return true;
  }
  if (value.length > 64) {
    return true;
  }
  return UNSAFE_VALUE_RE.test(value);
}

export function assertNoForbiddenKeys(payload: Record<string, unknown>): void {
  for (const key of Object.keys(payload)) {
    if (!isApprovedFieldName(key) || isForbiddenFieldName(key)) {
      throw new Error("unknown_or_forbidden_field");
    }
  }
}

File: engine/src/codestrata/telemetry/privacy.py
"""Structural privacy filtering for telemetry runtime events (Slice 9.1)."""

from __future__ import annotations

from codestrata.telemetry.errors import TelemetryRuntimeError, TelemetryRuntimeErrorCode

# Exact field-name denylist (authority). Safe categorical enum values that
# happen to contain substrings like "token" or "source" are evaluated as
# bounded enums, not via substring rejection of values.
FORBIDDEN_FIELD_NAMES: frozenset[str] = frozenset(
    {
        "repository",
        "repository_name",
        "repository_url",
        "project",
        "project_name",
        "organization",
        "source",
        "source_code",
        "code",
        "finding",
        "findings",
        "evidence",
        "recommendation",
        "recommendations",
        "path",
        "file",
        "filepath",
        "filename",
        "file_name",
        "directory",
        "cwd",
        "argv",
        "command_line",
        "command",
        "exception",
        "exception_message",
        "traceback",
        "stack",
        "stack_trace",
        "environment",
        "env",
        "credential",
        "credentials",
        "secret",
        "token",
        "password",
        "api_key",
        "authorization",
        "cookie",
        "prompt",
        "response",
        "model_id",
        "cost",
        "email",
        "user",
        "username",
        "account",
        "customer",
        "installation_id",
        "hostname",
        "home",
        "pid",
        "ip",
        "commit",
        "branch",
        "config_path",
        "output_path",
        "report_path",
        "package_name",
        "package_names",
        "exact_token_count",
        "token_count",
    }
)

APPROVED_FIELD_NAMES: frozenset[str] = frozenset(
    {
        "event_type",
        "client_name",
        "cli_version",
        "os_family",
        "arch_family",
        "lifecycle",
        "result",
        "duration_bucket",
        "operation_category",
        "enabled_assessment_heads",
        "offline_mode",
        "ai_used",
        "failure_category",
        "schema_version",
        "runtime_policy_version",
    }
)


def assert_field_name_allowed(name: str) -> None:
    if name in FORBIDDEN_FIELD_NAMES:
        raise TelemetryRuntimeError(TelemetryRuntimeErrorCode.UNSAFE_FIELD)
    if name not in APPROVED_FIELD_NAMES:
        raise TelemetryRuntimeError(TelemetryRuntimeErrorCode.UNKNOWN_FIELD)


def looks_like_path_or_secret_value(value: str) -> bool:
    """Conservative detector for accidental path/secret-shaped strings."""

    lowered = value.lower()
    if any(marker in value for marker in ("/", "\\", "://")):
        return True
    if any(
        marker in lowered
        for marker in (
            "-----begin",
            "sk-",
            "password=",
            "secret=",
            "api_key=",
            "bearer ",
        )
    ):
        return True
    return False


__all__ = [
    "APPROVED_FIELD_NAMES",
    "FORBIDDEN_FIELD_NAMES",
    "assert_field_name_allowed",
    "looks_like_path_or_secret_value",
]
